src/poly/multiopen/verifier.rs

```diff
@@ -0,0 +1,246 @@
+use super::super::{
+    commitment::{Guard, Params, MSM},
+    Error,
+};
+use super::{Proof, VerifierQuery};
+use crate::arithmetic::{
+    eval_polynomial, get_challenge_scalar, lagrange_interpolate, Challenge, CurveAffine, Field,
+};
+use crate::plonk::hash_point;
+use crate::transcript::Hasher;
+use std::collections::{BTreeMap, BTreeSet};
+
+#[derive(Debug, Clone)]
+struct CommitmentData<C: CurveAffine> {
+    set_index: usize,
+    point_indices: Vec<usize>,
+    evals: Vec<C::Scalar>,
+}
+
+impl<'a, C: CurveAffine> Proof<C> {
+    /// Verify a multi-opening proof
+    pub fn verify<I, HBase: Hasher<C::Base>, HScalar: Hasher<C::Scalar>>(
+        &self,
+        params: &'a Params<C>,
+        transcript: &mut HBase,
+        transcript_scalar: &mut HScalar,
+        queries: I,
+        msm: MSM<'a, C>,
+    ) -> Result<Guard<'a, C>, Error>
+    where
+        I: IntoIterator<Item = VerifierQuery<'a, C>> + Clone,
+    {
+        // Sample x_4 for compressing openings at the same points together
+        let x_4: C::Scalar = get_challenge_scalar(Challenge(transcript.squeeze().get_lower_128()));
+
+        let (commitment_map, point_sets) = construct_intermediate_sets::<'a, C, I>(queries.clone());
+
+        // Compress the commitments and expected evaluations at x_3 together.
+        // using the challenge x_4
+        let mut q_commitments: Vec<_> = vec![params.empty_msm(); point_sets.len()];
```

Comment:
I think we can do something here very similar to the `fold` stuff we did elsewhere.

Reply:
i do think we need to allocate `q_commitments` and `q_eval_sets` prior to accumulating their values. this is because each commitment is not being accumulated in order, but rather at an arbitrary `set_idx`.